### PR TITLE
[SLDEV-4588] Control log configuration from CLI commands

### PR DESCRIPTION
**Pull request by RoniSegal**

### CODE REVIEW

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class LogConfiguration {
+    public static final String SL_LOG_LEVEL = "sl.log.level";
+    public static final String SL_LOG_TO_CONSOLE = "sl.log.toConsole";
+    public static final String SL_LOG_TO_FILE = "sl.log.toFile";
+    public static final String SL_LOG_FOLDER = "sl.log.folder";
+    public static final String SL_LOG_FILE_NAME = "sl.log.filename";
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java:21]: You aren't referring this one at all.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 2]
```diff
@@ -33,14 +33,21 @@
     public boolean failBuildIfStepFail;
     public CommandMode commandMode;
     public CLIRunner cliRunner;
+    private String logFolder;
+    private LogDestination logDestination = LogDestination.CONSOLE;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java:37]: These will be the defaults for a user that upgrades the plugin?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 3]
```diff
@@ -132,6 +139,31 @@ public void setFailBuildIfStepFail(boolean failBuildIfStepFail) {
...
+
+    @Exported
+    public String getLogFolder() {
+        return logFolder;
+    }
+
+    @Exported
+    public LogLevel getLogLevel() {
+        return logLevel;
+    }
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java:160]: No reference to log filename.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java [hunk 4]
```diff
@@ -145,8 +177,8 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
 
             CLIHandler cliHandler = new CLIHandler(logger);
             cliRunner = createCLIRunner(commandMode);
-
-            isStepSuccessful = cliRunner.perform(build, launcher, listener, commandMode, cliHandler, logger);
+            LogConfiguration logConfiguration = new LogConfiguration(logFolder, logDestination, logLevel);
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java:180]: I suggest to create it one line before as that's in the middle of a flow.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/BaseCommandArguments.java [hunk 5]
```diff
@@ -32,142 +35,7 @@
     private AbstractBuild<?, ?> build;
     private EnvVars envVars;
     private Logger logger;
-
-    public String getAppName() {
-        return appName;
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/BaseCommandArguments.java:37]: Why did you remove all of these params? Only due to lombok or something else?

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/AbstractCommandExecutor.java [hunk 6]
```diff
@@ -21,71 +20,42 @@
...
-
-            printStreams(process);
-
-            if (process.exitValue() == 0) {
-                return true;
-            }
-
+            ProcessResult result = processExecutor.command(createExecutionCommand()).readOutput(true).execute();
+            String output = result.outputUTF8();
+            logger.info(output);
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/AbstractCommandExecutor.java:35]: I would like to see in the demo if that's easy to understand that this is the log of the step. Maybe it is worth prefixing it.
Also, make sure to test failure to reach the server and verify that the log is printed correctly.

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/ConfigTest.java [hunk 7]
```diff
@@ -98,12 +101,37 @@ protected BaseCommandArguments createBaseCommandArguments() throws IOException {
...
         baseCommandArguments.setBuild(build);
 
         return baseCommandArguments;
     }
+
+    private List<String> createExpectedCommand(BaseCommandArguments baseCommandArguments) {
+        List<String> expected = new ArrayList<>();
+        expected.add("path/to/java");
+        expected.add(AbstractCommandExecutor.formatTagProp());
+        expected.addAll(baseCommandArguments.getLogConfiguration().toSystemProperties());
```
Comment (nadavye) [inline src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/ConfigTest.java:116]: Please test explicitly the field names.

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/SendExternalReportTest.java [hunk 8]
```diff
@@ -80,7 +82,28 @@ private BaseCommandArguments createBaseCommandArguments() {
...
         return baseCommandArguments;
     }
 
+    private List<String> createExpectedCommand(BaseCommandArguments baseCommandArguments) {
+        List<String> expected = new ArrayList<>();
+        expected.add("path/to/java");
+        expected.add(AbstractCommandExecutor.formatTagProp());
+        expected.addAll(baseCommandArguments.getLogConfiguration().toSystemProperties());
+        expected.add("-jar");
+        expected.add("agent.jar");
```
Comment (nadavye) [inline src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/SendExternalReportTest.java:95]: This code is very similar in all of the tests.
Please refactor.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java (hunk 1 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java:21]: Done

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java (hunk 3 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java:162]: Done

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/BaseCommandArguments.java (hunk 5 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/entities/BaseCommandArguments.java:37]: Just for lombok

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/ConfigTest.java (hunk 7 above)
Comment (RoniSegal) [reply src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/ConfigTest.java:116]: Here? this is test class for config command, the real values are tested in `LogConfiurationTest` (all the permutations)

File: src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/SendExternalReportTest.java (hunk 8 above)
Comment (RoniSegal) [reply src/test/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/executors/SendExternalReportTest.java:95]: Absolutely agree, but i need to do some hard refactoring (probably using a base class).
Dont think its part of this scope.
I opened  a ticket: https://sealights.atlassian.net/browse/SLDEV-7231

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java (hunk 4 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java]: Done

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java (hunk 2 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep.java:38]: Yes

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java [hunk 9]
```diff
@@ -41,7 +48,9 @@
                 properties.add(formatKeyValue(SL_LOG_FILE_NAME, logFilename));
             }
             if (!StringUtils.isNullOrEmpty(logFolder)) {
-                properties.add(formatKeyValue(SL_LOG_FOLDER, logFolder));
+                String resolvedLogFolder = resolveLogFolder(logFolder);
+                logger.info("Log folder set to " + resolvedLogFolder);
```
Comment (nadavye) [inline src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java:52]: Please quote the value of `resolvedLogFolder`

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly [hunk 10]
```diff
@@ -40,8 +40,8 @@
             <f:enum>${it.displayName}</f:enum>
         </f:entry>
 
-        <f:entry title="Logs Folder" field="logFolder" description="The path in which the logs will be saved.">
-            <f:textbox value="${instance.logFolder}"/>
+        <f:entry title="Logs Folder" field="logFolder" description="The path in which the logs will be saved. (relative to workspace)">
+            <f:textbox value="${instance.logFolder}" default="$${WORKSPACE}/sl-log"/>
```
Comment (nadavye) [inline src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly:44]: Please set the default to `sealights-logs` and make sure that the filename will be `jenkins-cli.log`.

File: src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java (hunk 9 above)
Comment (RoniSegal) [reply src/main/java/io/sealights/plugins/sealightsjenkins/buildsteps/cli/LogConfiguration.java:52]: Done

File: src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly (hunk 10 above)
Comment (RoniSegal) [reply src/main/resources/io/sealights/plugins/sealightsjenkins/buildsteps/cli/SealightsCLIBuildStep/config.jelly:44]: Per the folder name - done.
Per the file name, the logs are coming from the agent which named `sl-java-agent.log`, more clear than jenkins-cli
